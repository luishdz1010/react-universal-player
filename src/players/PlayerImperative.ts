export default interface PlayerImperative<T> {
  readonly player: T

  seekTo(amount: number): void
}
