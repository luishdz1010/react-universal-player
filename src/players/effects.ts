import { useEffect } from 'react'
import PlayerImperativePrivate, { PlayerImperativeMuteable } from './private/PlayerImperativePrivate'

export const usePlaybackEffect = (player: PlayerImperativePrivate | null, playing: boolean) => {
  useEffect(() => {
    if (playing) player?.play()
    else player?.pause()
  }, [playing, player])
}

export const useLoopEffect = (player: PlayerImperativePrivate | null, loop: boolean) => {
  useEffect(() => {
    player?.setLoop(loop)
  }, [loop, player])
}

export interface Volume {
  volume: number
  muted: boolean
}

export const useVolumeMutedEffect = (
  player: (PlayerImperativePrivate & PlayerImperativeMuteable) | null,
  { muted, volume }: Volume
) => {
  useEffect(() => {
    player?.setVolume(volume)
  }, [player, volume])

  useEffect(() => {
    player?.setMuted(muted)
  }, [player, muted])
}

export const useVolumeMutedUnifiedEffect = (player: PlayerImperativePrivate | null, { muted, volume }: Volume) => {
  useEffect(() => {
    if (!muted) player?.setVolume(volume)
  }, [muted, player, volume])

  useEffect(() => {
    if (muted) player?.setVolume(0)
  }, [muted, player])
}
