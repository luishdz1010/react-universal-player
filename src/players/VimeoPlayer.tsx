import { Options } from '@vimeo/player'
import React, { forwardRef, useEffect, useImperativeHandle, useRef, useState } from 'react'
import { PlayerProps, VideoState } from '../PlayerProps'
import { vimeoUrlTest, VimeoVendorApi } from '../providers/vimeo'
import useVideoState from '../useVideoState'
import { useLoopEffect, usePlaybackEffect, useVolumeMutedUnifiedEffect } from './effects'
import useCreatePlayer from './private/useCreatePlayer'
import { useLastValueRef, useValidateUrl } from './private/util'
import VimeoIframe, { VimeoIframeProps } from './VimeoIframe'
import VimeoPlayerImpl, { VimeoPlayerImperative } from './private/VimeoPlayerImpl'

export interface VimeoPlayerProps extends PlayerProps {
  vimeoVendorApi?: VimeoVendorApi | null
  vimeoVendorApiLoader?: () => Promise<VimeoVendorApi>
  vimeoColor?: string
  vimeoOptions?: Options
  vimeoIframeProps?: VimeoIframeProps
}

export interface VimeoOwnPlayerState {
  vimeoColor: string | null
  vimeoOptions: Options
}

const useVimeoOwnState = (args: VimeoPlayerProps): VimeoOwnPlayerState => {
  return {
    vimeoColor: args.vimeoColor || null,
    vimeoOptions: args.vimeoOptions || {},
  }
}

export interface VimeoPlayerState extends VideoState<VimeoVendorApi>, VimeoOwnPlayerState {}

export const useVimeoPlayerState = (args: VimeoPlayerProps): VimeoPlayerState => ({
  ...useVideoState(args, {
    vendorApi: args.vimeoVendorApi,
    vendorApiLoader: args.vimeoVendorApiLoader,
  }),
  ...useVimeoOwnState(args),
})

const extractVideoId = (url: string) => Number(url.match(/(\d+)\/?(?:[#?].*)?$/)![1])

const playerLifecycle = {
  create(stateRef: { current: VimeoPlayerState }, rootEl: HTMLIFrameElement) {
    const init = stateRef.current
    const Vimeo = init.vendorApi!
    const id = extractVideoId(init.url)

    const raw = new Vimeo.Player(rootEl, {
      id,
      autoplay: init.playing,
      controls: init.controls,
      playsinline: init.playsinline,
      loop: init.loop,
      muted: init.muted,
      ...init.vimeoOptions,
    })

    const api = new VimeoPlayerImpl(raw, id)

    raw.on('play', () => {
      stateRef.current.onPlayingChange(true)
    })

    raw.on('pause', () => {
      stateRef.current.onPlayingChange(false)
    })

    raw.on('volumechange', ({ volume }: { volume: number }) => {
      stateRef.current.onVolumeChange(volume)
    })

    return api
  },

  destroy(api: VimeoPlayerImpl) {
    api.player.off('play')
    api.player.off('pause')
    api.player.off('volumechange')
    api.player.unload()
  },
}

const VimeoPlayer = forwardRef<VimeoPlayerImperative | null, VimeoPlayerProps>((props, outerRef) => {
  const state = useVimeoPlayerState(props)
  const rootEl = useRef<HTMLIFrameElement>(null)

  if (typeof window !== 'undefined') {
    const stateRef = useLastValueRef(state)

    const { url, vendorApi, controls, playsinline, vimeoOptions, loop, playing, volume, muted, vimeoColor } = state

    const [recreate, recreatePlayer] = useState({})

    const isValidUrl = useValidateUrl(url, vimeoUrlTest, stateRef.current.onError)

    const player = useCreatePlayer(
      {
        shouldExist: isValidUrl && !!vendorApi,
        create: () => playerLifecycle.create(stateRef, rootEl.current!),
        destroy: playerLifecycle.destroy,
        onError: stateRef,
      },
      [controls, playsinline, recreate, ...Object.values(vimeoOptions)]
    )

    usePlaybackEffect(player, playing)

    useLoopEffect(player, loop)

    useVolumeMutedUnifiedEffect(player, { muted, volume })

    useEffect(() => {
      let stale = false

      if (isValidUrl)
        player?.loadVideo(extractVideoId(url)).catch((err: Error) => {
          if (!stale) stateRef.current.onError(err)
        })

      return () => {
        stale = true
      }
    }, [isValidUrl, url, player, stateRef])

    useEffect(() => {
      if (vimeoColor) player?.setColor(vimeoColor)
    }, [vimeoColor, player])

    useEffect(() => {
      if (!vimeoColor) recreatePlayer({})
    }, [vimeoColor])

    useImperativeHandle(outerRef, () => player, [player])
  }

  const { current: init } = useRef(state)
  const iframeProps = useRef(props.vimeoIframeProps)

  return (
    <VimeoIframe
      className={props.className}
      urlConstructor={init.urlConstructor}
      url={init.url}
      playing={init.playing}
      controls={init.controls}
      loop={init.loop}
      muted={init.muted}
      playsinline={init.playsinline}
      vimeoOptions={init.vimeoOptions}
      {...iframeProps.current}
      ref={rootEl}
    />
  )
})

export default VimeoPlayer
