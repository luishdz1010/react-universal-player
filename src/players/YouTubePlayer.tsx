import React, { ComponentProps, forwardRef, useImperativeHandle, useMemo, useRef } from 'react'
import { PlayerProps, VideoState } from '../PlayerProps'
import { youTubeUrlTest, YouTubeVendorApi } from '../providers/youtube'
import useVideoState from '../useVideoState'
import { useLoopEffect, usePlaybackEffect, useVolumeMutedEffect } from './effects'
import useCreatePlayer from './private/useCreatePlayer'
import { useLastValueRef, useValidateUrl } from './private/util'
import YouTubePlayerImpl, { YouTubePlayerImperative } from './private/YouTubePlayerImpl'

export interface YouTubePlayerProps extends PlayerProps {
  youTubeVendorApi?: YouTubeVendorApi | null
  youTubeVendorApiLoader?: () => Promise<YouTubeVendorApi>
  youTubePlayerOptions?: YT.PlayerOptions
  youTubePlayerVars?: YT.PlayerVars
  youTubeIframeProps?: ComponentProps<'iframe'>
}

export interface YouTubeOwnPlayerState {
  youTubePlayerOptions: Omit<YT.PlayerOptions, 'playerVars' | 'events'>
  youTubePlayerVars: YT.PlayerVars
}

const useYouTubeOwnState = (args: YouTubePlayerProps): YouTubeOwnPlayerState => {
  return {
    youTubePlayerOptions: args.youTubePlayerOptions || {},
    youTubePlayerVars: args.youTubePlayerVars || {},
  }
}

export interface YouTubePlayerState extends VideoState<YouTubeVendorApi>, YouTubeOwnPlayerState {}

export const useYouTubePlayerState = (args: YouTubePlayerProps): YouTubePlayerState => ({
  ...useVideoState(args, {
    vendorApi: args.youTubeVendorApi,
    vendorApiLoader: args.youTubeVendorApiLoader,
  }),
  ...useYouTubeOwnState(args),
})

const extractVideoId = (url: string) => {
  const match = url.match(/(?:youtu\.be\/|youtube\.com\/(?:embed\/|v\/|watch\?v=|watch\?.+&v=))(?:\w|-)+/)

  return (match && match[1]) || undefined
}

const playerLifecycle = {
  create(
    state: { current: YouTubePlayerState },
    rootEl: HTMLIFrameElement
  ): [Promise<YouTubePlayerImpl | null>, () => void] {
    let stale = false

    const promise = new Promise<YouTubePlayerImpl | null>((accept, reject) => {
      const init = state.current
      const YT = init.vendorApi!

      // not documented on YouTube API
      const mute = init.muted ? { mute: 1 } : null

      const raw = new YT.Player(rootEl, {
        videoId: extractVideoId(init.url),
        ...init.youTubePlayerOptions,
        playerVars: {
          origin: window.location.origin,
          autoplay: init.playing ? YT.AutoPlay.AutoPlay : YT.AutoPlay.NoAutoPlay,
          controls: init.controls ? YT.Controls.ShowLoadPlayer : YT.Controls.Hide,
          playsinline: init.playsinline ? YT.PlaysInline.Inline : YT.PlaysInline.Fullscreen,
          ...mute,
          ...init.youTubePlayerVars,
        },
        events: {
          onReady: () => {
            if (!stale) {
              raw.addEventListener('onStateChange', onStateChange)
              const api = new YouTubePlayerImpl(raw, onStateChange)
              accept(api)
            } else {
              accept(null)
            }
          },
          onError(event) {
            reject(new Error(`YouTube player error ${event.data}`))
          },
        },
      })

      function onStateChange(ev: YT.OnStateChangeEvent) {
        switch (ev.data) {
          case YT.PlayerState.BUFFERING:
          case YT.PlayerState.PLAYING:
            if (!state.current.playing) {
              state.current.onPlayingChange(true)
            }
            break

          case YT.PlayerState.PAUSED:
          case YT.PlayerState.ENDED:
            if (state.current.playing) {
              state.current.onPlayingChange(false)
            }
            break

          case YT.PlayerState.UNSTARTED:
          case YT.PlayerState.CUED:
          default:
            break
        }
      }
    })

    return [
      promise,
      () => {
        stale = true
      },
    ]
  },

  destroy(api: YouTubePlayerImpl) {
    api.player.removeEventListener('onStateChange', api.onStateChange)
    api.player.stopVideo()
  },
}

const YouTubePlayer = forwardRef<YouTubePlayerImperative | null, YouTubePlayerProps>((props, outerRef) => {
  const state = useYouTubePlayerState(props)
  const rootEl = useRef<HTMLIFrameElement>(null)

  if (typeof window !== 'undefined') {
    const stateRef = useLastValueRef(state)

    const {
      url,
      vendorApi,
      controls,
      playsinline,
      loop,
      playing,
      volume,
      muted,
      youTubePlayerOptions,
      youTubePlayerVars,
    } = state

    const isValidUrl = useValidateUrl(url, youTubeUrlTest, stateRef.current.onError)

    const player = useCreatePlayer(
      {
        shouldExist: isValidUrl && !!vendorApi,
        create: () => playerLifecycle.create(stateRef, rootEl.current!),
        destroy: playerLifecycle.destroy,
        onError: stateRef,
      },
      [url, controls, playsinline, ...Object.values(youTubePlayerOptions), '␃', ...Object.values(youTubePlayerVars)]
    )

    usePlaybackEffect(player, playing)

    useLoopEffect(player, loop)

    useVolumeMutedEffect(player, { muted, volume })

    useImperativeHandle(outerRef, () => player, [player])
  }

  const initialStateRef = useRef(state)

  const iframeUrl = useMemo(() => {
    const init = initialStateRef.current

    try {
      const URL = init.urlConstructor
      const u = new URL(init.url)
      const T = '1'
      const F = '0'

      u.searchParams.set('autoplay', init.playing ? T : F)
      u.searchParams.set('loop', init.loop ? T : F)
      u.searchParams.set('playsinline', init.playsinline ? T : F)
      u.searchParams.set('controls', init.controls ? T : F)
      u.searchParams.set('mute', init.muted ? T : F) // not documented

      // see https://developers.google.com/youtube/player_parameters#loop
      // u.searchParams.set('playlist', videoId)

      u.searchParams.set('modestbranding', T)
      u.searchParams.set('showinfo', F)
      u.searchParams.set('rel', F)
      u.searchParams.set('enablejsapi', T)
      u.searchParams.set('iv_load_policy', T)

      return u.toString()
    } catch {
      return init.url
    }
  }, [])

  const iframeProps = useRef(props.youTubeIframeProps)

  return (
    <iframe
      className={props.className}
      ref={rootEl}
      src={iframeUrl}
      allowFullScreen
      allow="autoplay; fullscreen"
      frameBorder="0"
      {...iframeProps.current}
    />
  )
})

export default YouTubePlayer
