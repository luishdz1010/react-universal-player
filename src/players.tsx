import React from 'react'
import { PlayerProps } from './PlayerProps'
import VimeoPlayer from './player/VimeoPlayer'
import YouTubePlayer from './player/YouTubePlayer'
import { vimeoUrlTest, vimeoVendorApiLoader } from './providers/vimeo'
import { youTubeUrlTest, youTubeVendorApiLoader } from './providers/youtube'

interface PlayerDef {
  test: (url: string) => boolean
  render: (props: PlayerProps) => React.ReactElement
}

const players: PlayerDef[] = [
  {
    test: youTubeUrlTest,
    render(props) {
      return <YouTubePlayer youTubeVendorApiLoader={youTubeVendorApiLoader} {...props} />
    },
  },
  {
    test: vimeoUrlTest,
    render(props) {
      return <VimeoPlayer vimeoVendorApiLoader={vimeoVendorApiLoader} {...props} />
    },
  },
]

export default players
