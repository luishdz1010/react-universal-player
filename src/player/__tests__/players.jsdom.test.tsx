import { render, wait } from '@testing-library/react'
import React from 'react'
import { describe, expect, it } from '../../__tests__/jest'
import * as allPlayers from './players'
import { commonPlayersTests } from './players.jest'

commonPlayersTests(URL)

describe.each(Object.values(allPlayers))('%s jsdom', ({ Player }) => {
  const sharedPlayerProps = {
    urlConstructor: URL,
  }

  it('returns an error unrecognizable url', async () => {
    let err: Error | null = null

    render(
      <Player
        url="http://example.com/123"
        {...sharedPlayerProps}
        onPlayingChange={({ error }) => {
          err = error
        }}
      />
    )

    await wait(() => {
      expect(err).toBeInstanceOf(Error)
      expect(err).toMatchSnapshot()
    })
  })
})
