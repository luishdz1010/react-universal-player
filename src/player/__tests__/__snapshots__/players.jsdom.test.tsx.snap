// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VimeoPlayer jsdom calls onError on unrecognizable url 1`] = `[Error: provided url(http://example.com/123) isn't supported by this player]`;

exports[`VimeoPlayer jsdom calls onPlaybackError on unrecognizable url 1`] = `[Error: provided url(http://example.com/123) isn't supported by this player]`;

exports[`VimeoPlayer jsdom returns an error unrecognizable url 1`] = `[Error: provided url(http://example.com/123) isn't supported by this player]`;

exports[`VimeoPlayer universal renders with url= 1`] = `
<DocumentFragment>
  <iframe
    allow="autoplay; fullscreen"
    allowfullscreen=""
    frameborder="0"
    src=""
  />
</DocumentFragment>
`;

exports[`VimeoPlayer universal renders with url=http://example.com 1`] = `
<DocumentFragment>
  <iframe
    allow="autoplay; fullscreen"
    allowfullscreen=""
    frameborder="0"
    src="http://example.com/?autoplay=0&controls=1&loop=0&muted=0&playsinline=0"
  />
</DocumentFragment>
`;

exports[`VimeoPlayer universal renders with url=notaurl 1`] = `
<DocumentFragment>
  <iframe
    allow="autoplay; fullscreen"
    allowfullscreen=""
    frameborder="0"
    src="notaurl"
  />
</DocumentFragment>
`;

exports[`VimeoPlayer universal renders with url=video.com 1`] = `
<DocumentFragment>
  <iframe
    allow="autoplay; fullscreen"
    allowfullscreen=""
    frameborder="0"
    src="video.com"
  />
</DocumentFragment>
`;

exports[`YouTubePlayer jsdom calls onError on unrecognizable url 1`] = `[Error: provided url(http://example.com/123) isn't supported by this player]`;

exports[`YouTubePlayer jsdom returns an error unrecognizable url 1`] = `[Error: provided url(http://example.com/123) isn't supported by this player]`;

exports[`YouTubePlayer universal renders with url= 1`] = `
<DocumentFragment>
  <iframe
    allow="autoplay; fullscreen"
    allowfullscreen=""
    frameborder="0"
    src=""
  />
</DocumentFragment>
`;

exports[`YouTubePlayer universal renders with url=http://example.com 1`] = `
<DocumentFragment>
  <iframe
    allow="autoplay; fullscreen"
    allowfullscreen=""
    frameborder="0"
    src="http://example.com/?autoplay=0&loop=0&playsinline=0&controls=1&mute=0&modestbranding=1&showinfo=0&rel=0&enablejsapi=1&iv_load_policy=1"
  />
</DocumentFragment>
`;

exports[`YouTubePlayer universal renders with url=notaurl 1`] = `
<DocumentFragment>
  <iframe
    allow="autoplay; fullscreen"
    allowfullscreen=""
    frameborder="0"
    src="notaurl"
  />
</DocumentFragment>
`;

exports[`YouTubePlayer universal renders with url=video.com 1`] = `
<DocumentFragment>
  <iframe
    allow="autoplay; fullscreen"
    allowfullscreen=""
    frameborder="0"
    src="video.com"
  />
</DocumentFragment>
`;
